sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala

```diff
@@ -128,6 +128,97 @@ private[sql] object ArrowConverters extends Logging {
     }
   }
 
+  private[sql] def toArrowBatchIterator(
+      rowIter: Iterator[InternalRow],
+      schema: StructType,
+      maxRecordsPerBatch: Int,
+      timeZoneId: String): Iterator[(Array[Byte], Long, Long)] = {
+    val arrowSchema = ArrowUtils.toArrowSchema(schema, timeZoneId)
+    val allocator = ArrowUtils.rootAllocator.newChildAllocator(
+      "toArrowBatchIterator", 0, Long.MaxValue)
+
+    val root = VectorSchemaRoot.create(arrowSchema, allocator)
+    val unloader = new VectorUnloader(root)
+    val arrowWriter = ArrowWriter.create(root)
+
+    Option(TaskContext.get).foreach {
+      _.addTaskCompletionListener[Unit] { _ =>
+        root.close()
+        allocator.close()
+      }
+    }
+
+    new Iterator[(Array[Byte], Long, Long)] {
+
+      override def hasNext: Boolean = rowIter.hasNext || {
+        root.close()
+        allocator.close()
+        false
+      }
+
+      override def next(): (Array[Byte], Long, Long) = {
+        val out = new ByteArrayOutputStream()
+        val writeChannel = new WriteChannel(Channels.newChannel(out))
+
+        var rowCount = 0L
+        var estimatedSize = SizeEstimator.estimate(arrowSchema) +
+          SizeEstimator.estimate(IpcOption.DEFAULT)
+        Utils.tryWithSafeFinally {
+          while (rowIter.hasNext && (maxRecordsPerBatch <= 0 || rowCount < maxRecordsPerBatch)) {
+            val row = rowIter.next()
+            arrowWriter.write(row)
+            rowCount += 1
+            estimatedSize += SizeEstimator.estimate(row)
+          }
+          arrowWriter.finish()
+          val batch = unloader.getRecordBatch()
+
+          MessageSerializer.serialize(writeChannel, arrowSchema)
+          MessageSerializer.serialize(writeChannel, batch)
+          ArrowStreamWriter.writeEndOfStream(writeChannel, IpcOption.DEFAULT)
+
+          batch.close()
+        } {
+          arrowWriter.reset()
+        }
+
+        (out.toByteArray, rowCount, estimatedSize)
+      }
+    }
+  }
+
+  private[sql] def createEmptyArrowBatch(
```

Comment:
Can we deduplicate the codes with `toBatchIterator`? e.g.) you can:

```scala
private[sql] def createArrowBatch(
    handleBatch: RecordBatch => Unit,
    ...) = {
  ...
  handleBatch(unloader.getRecordBatch())
  ...
}

def toArrowBatchIterator = {
  ...
  createArrowBatch(..., { batch =>
    ...
  }
  ...
}

def toBatchIterator = {
  ...
  createArrowBatch(..., { batch =>
    ...
  }
  ...
}
```

Reply:
let me take a look

Reply:
still don't figure out how to deduplicate, the `toArrowBatchIterator` should also return `rowCount`

what about trying to do this after switch to `batch size < 4MB`?

Reply:
calling toArrowBatchIterator with an empty iterator will return an empty iterator

here needs an arrow batch with empty data